frontend/document/credocument.lua

```diff
@@ -283,10 +287,127 @@ function CreDocument:updateColorRendering()
     end
 end
 
+function CreDocument:setHideNonlinear(hide_nonlinear)
+    if hide_nonlinear ~= self.hide_nonlinear and self._document:hasNonLinearFlows() then
+        self.hide_nonlinear = hide_nonlinear
+        self._document:setIntProperty("crengine.doc.nonlinear.pagebreak.force", self.hide_nonlinear and 1 or 0)
+        self:cacheFlows()
+    end
+end
+
 function CreDocument:getPageCount()
     return self._document:getPages()
 end
 
+-- Get the next/prev page number, skipping non-linear flows,
+-- i.e. the next/prev page that is either in the current
+-- flow or in the linear flow (flow 0)
+-- If "page" is 0, these give the initial and final linear pages
+function CreDocument:getNextPage(page)
+    if self:hasFlows() then
+        if page < 0 then return 0 end
+        local flow = page > 0 and self:getPageFlow(page) or 0
+        for test_page=page+1, self:getPageCount() do
+            local test_page_flow = self:getPageFlow(test_page)
+            if test_page_flow == flow or test_page_flow == 0 then
+                return test_page
+            end
+        end
+        return 0
+    else
+        return Document.getNextPage(self, page)
+    end
+end
+
+function CreDocument:getPrevPage(page)
+    if self:hasFlows() then
+        if page > self:getPageCount()+1 then return 0 end
+        local flow = page > 0 and self:getPageFlow(page) or 0
+        for test_page=(page==0) and self:getPageCount() or page-1, 1, -1 do
+            local test_page_flow = self:getPageFlow(test_page)
+            if test_page_flow == flow or test_page_flow == 0 then
+                return test_page
+            end
+        end
+        return 0
+    else
+        return Document.getPrevPage(self, page)
+    end
+end
+
+function CreDocument:getPageFlow(page)
+    -- Only report non-linear pages if "hide_nonlinear" is enabled, and in 1-page mode,
+    -- otherwise all pages are linear (flow 0)
+    if self.hide_nonlinear and self._view_mode == self.PAGE_VIEW_MODE and self:getVisiblePageCount() == 1 then
+        return self._document:getPageFlow(page)
+    else
+        return 0
+    end
+end
+
+function CreDocument:getFirstInFlow(flow)
+    return self.flows[flow][1]
+end
+
+function CreDocument:getTotalPagesInFlow(flow)
+    return self.flows[flow][2]
+end
+
+function CreDocument:getPageNumberInFlow(page)
+    if self:hasFlows() then
+        return self.page_in_flow[page]
+    else
+        return page
+    end
+end
+
+function CreDocument:hasFlows()
```

Comment:
I think just seeing this one named `hasHiddenFlows()` and checking for `self.hide_nonlinear_flows` - and abused a bit more as wrappers - would make me quite happy with this PR.
(And moved above before all the other` :*Flow*` functions.)